File: python/problems/leetcode/0237_delete_node_in_a_linked_list/solution.py
"""LeetCode 237: Delete Node in a Linked List."""

from __future__ import annotations

from typing import Optional


class ListNode:
    def __init__(self, val: int = 0, next: Optional["ListNode"] = None) -> None:
        self.val = val
        self.next = next


class Solution:
    def deleteNode(self, node: ListNode) -> None:
        assert node.next is not None
        node.val = node.next.val
        node.next = node.next.next


def to_list(h: Optional[ListNode]) -> list[int]:
    out: list[int] = []
    while h is not None:
        out.append(h.val)
        h = h.next
    return out


if __name__ == "__main__":
    a = ListNode(4)
    b = ListNode(5)
    c = ListNode(1)
    d = ListNode(9)
    a.next = b
    b.next = c
    c.next = d
    Solution().deleteNode(b)
    assert to_list(a) == [4, 1, 9]
    print("0237 OK")

File: cpp/problems/leetcode/0237_delete_node_in_a_linked_list/solution.cpp
// LeetCode 237
#include <bits/stdc++.h>
#include <cassert>
using namespace std;

struct ListNode {
    int val;
    ListNode* next;
    ListNode(int v) : val(v), next(nullptr) {}
};

class Solution {
public:
    void deleteNode(ListNode* node) {
        node->val = node->next->val;
        node->next = node->next->next;
    }
};

int main() {
    ListNode a(4), b(5), c(1), d(9);
    a.next = &b;
    b.next = &c;
    c.next = &d;
    Solution().deleteNode(&b);
    vector<int> out;
    for (ListNode* p = &a; p; p = p->next) out.push_back(p->val);
    assert((out == vector<int>{4, 1, 9}));
    cout << "0237 OK" << endl;
    return 0;
}
